### PR TITLE
Allow to change transaction name

### PR DESCRIPTION
**Pull request by olerapx**
Add a PHP function to change transaction name from user code.

This is useful if a PHP project uses GraphQL API. In that case, all requests will be routed through a single endpoint (that is, /graphql), and it becomes impossible to group requests based on endpoint parameters, such as in the REST API case

### CODE REVIEW

File: src/request.rs [hunk 1]
```diff
@@ -332,3 +332,14 @@ fn finish_request_context(request_id: Option<i64>, status_code: i32) -> crate::R
 
     Ok(())
 }
+
+pub fn skywalking_set_request_name(args: &mut [ZVal]) -> phper::Result<()> {
```
Comment (wu-sheng) [inline src/request.rs:336]: `operation name` is the official term for the span. Why is this called `request name`? We don't have this concept.

Comment (olerapx) [reply src/request.rs]: Renamed to skywalking_set_operation_name

**Comment by olerapx** [comment]
> @olerapx Therefore, if you want to change the `operation_name` of `/graphql`, would it be better to build a GraphQL plug-in? This method is not restricted to GraphQL because its reach is too broad.

That would involve parsing the incoming GraphQL query, which feels out of scope for the project. I can, of course, use something like https://docs.rs/graphql-parser/latest/graphql_parser/ and extract operation name from the query, but I think it's more appropriate to leave for the usercode to handle.

The function can be applied not only for GraphQL, sure. That's the point: if the currently provided mechanism of constructing the operation name is not suitable for any particular project, it can be customized from PHP without the need of modifying the skywalking-php further :)

**Comment by wu-sheng** [comment]
The side effect of exposing and using this API in your app code, would make your app can't run w/o agent codes. Such as when you face performance impact of extra payload or/and in some circumstances.

**Comment by olerapx** [comment]
> The side effect of exposing and using this API in your app code, would make your app can't run w/o agent codes. Such as when you face performance impact of extra payload or/and in some circumstances.

That can be mitigated with a simple `if (extension_loaded('skywalking_agent'))` check in PHP code. Speaking from NewRelic/Atatus usage experience, it's basically the same with those APMs.

I think, it should just be mentioned somewhere in the docs.

**Comment by wu-sheng** [comment]
Generally, if you refer to our Java agent implementation(go agent are going to have this as well), https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/application-toolkit-tracer/, we recommend you to build a toolkit lib with empty implementation and make sure they are harmless.

Your codes should always use the toolkit lib without concerns, and the agent would instrument those APIs when the agent installed.

This is better to add many IF/ELSE in your app codes.

**Comment by olerapx** [comment]
Do you mean creating a PHP library wrapping the `skywalking_set_request_name` (and, potentially, other) calls with `extension_enabled` check?
I fully agree that's better than adding if/else in code everytime :)

**Comment by wu-sheng** [comment]
> Do you mean creating a PHP library wrapping the skywalking_set_request_name (and, potentially, other) calls with extension_enabled check?

You could check things in Java, https://github.com/apache/skywalking-java/blob/main/apm-application-toolkit/apm-toolkit-trace/src/main/java/org/apache/skywalking/apm/toolkit/trace/ActiveSpan.java
All these methods in the lib are empty, so, your app codes could use them at any place you like. 
When the extension(PHP agent) is installed, these methods would be enhanced by the agent, and then, the interceptor of these methods would call the original methods. Through this way, you actually don't need any `if/else`. You could install the agent or uninstall the agent at any time.

**Comment by jmjoy** [comment]
It is not recommended to use extension functions, as discussed above, when `skywalking_agent` extension is not loaded, it will result in fatal errors. My suggestion is to use environment variables or other system variables, such as `putenv` in user codes, and then obtain the environment variables in the extension for further processing.

**Comment by wu-sheng** [comment]
> It is not recommended to use extension functions, as discussed above, when `skywalking_agent` extension is not loaded, it will result in fatal errors. My suggestion is to use environment variables or other system variables, such as `putenv` in user codes, and then obtain the environment variables in the extension for further processing.

`setOperationName` is a runtime operation per request context, how could env work for this case?

**Comment by jmjoy** [comment]
> > It is not recommended to use extension functions, as discussed above, when `skywalking_agent` extension is not loaded, it will result in fatal errors. My suggestion is to use environment variables or other system variables, such as `putenv` in user codes, and then obtain the environment variables in the extension for further processing.
> 
> `setOperationName` is a runtime operation per request context, how could env work for this case?

For php-fpm, env is request context.

**Comment by wu-sheng** [comment]
OK, if that is the case, it should be good to use it as a bridge between app codes and agent kernel.

**Comment by wu-sheng** [comment]
Close as no active for months.

**Pull request closed without merging**